[SDK] Fix multiple issues with Swift KVO

lilyball opened the pull request:
This PR fixes a number of issues with Swift KVO. It fixes a bad thread safety issue, a problem where `NSKeyValueObservation` was incorrectly replacing `NSObject.observeValue(forKeyPath:of:change:context:)`, a potential race with `NSKeyValueObservation` handling a KVO notice during `deinit`, an issue with `NSSortDescriptor.keyPath` returning the wrong value, and a problem where the `NSKeyValueObservingCustomization` methods could be called with the wrong `KeyPath` if `KeyPath`-related work is happening concurrently on another thread.

Resolves [SR-6795](https://bugs.swift.org/browse/SR-6795), [SR-9074](https://bugs.swift.org/browse/SR-9074), [SR-9075](https://bugs.swift.org/browse/SR-9075), [SR-9076](https://bugs.swift.org/browse/SR-9076), and [SR-9077](https://bugs.swift.org/browse/SR-9077).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

lilyball committed:
[SDK] Fix thread safety of KVO swizzling

Swizzling the methods in a different order ensures we don't overwrite
the public `NSObject` method until it's safe to call our replacement.

rdar://problem/36663633 https://bugs.swift.org/browse/SR-6795

lilyball committed:
[SDK] Fix potential race with NSKeyValueObservation

If the `NSKeyValueObservation` was being deinited on one thread while a
KVO change notice was being broadcast on another, it could end up trying
to handle the change notice concurrently with deiniting, which will
probably crash.

This also fixes a problem where it was swizzling `NSObject`'s
`observeValue(forKeyPath:of:change:context:)` implementation, which
resulted in potentially invoking undefined behavior if an observer ever
called up to `NSObject`'s implementation.

https://bugs.swift.org/browse/SR-9074
https://bugs.swift.org/browse/SR-9075

lilyball committed:
[SDK] Fix NSSortDescriptor.keyPath

The previous implementation could return keyPaths other than the one
used to create the `NSSortDescriptor` in the event that any subsequent
observation, `NSExpression`, or `NSSortDescriptor` was created using a
different `KeyPath` that had the same Obj-C keypath value.

https://bugs.swift.org/browse/SR-9076

lilyball committed:
[SDK] Replace global KeyPath table with a per-thread table

The global table was vulnerable to race conditions when making
observations concurrently on multiple threads. We can assume the
`NSKeyValueObservingCustomization` methods are invoked synchronously
when creating the observation, so we can use a per-thread table instead.
This per-thread table is still vulnerable in the event that an
implementation of `automaticallyNotifiesObservers(for:)` creates a new
observation using a different `KeyPath` with the same Obj-C path, but
this method really shouldn't be creating any observations of its own.

Unfortunately we can't include the class in the per-thread table key, as
this may produce incorrect results in the presence of subclasses.

rdar://problem/45567020 https://bugs.swift.org/browse/SR-9077

millenomi commented:
@swift-ci please smoke test macos

millenomi commented:
One test is failing:

```
/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/stdlib/KVOKeyPaths.swift:107:16: error: CHECK-NEXT: expected string not found in input
```

kballard commented:
Fun fact: That test is actually broken. As in, the test was confirming that the previous behavior of this code was broken. The implementation of `automaticallyNotifiesObservers(for:)` was never actually being invoked with the `\Target.objcValue2` or `Target.objcValue3` key paths, because those key paths were never explicitly used prior and therefore never put into the global table. The code itself even has comments saying which lines are intended to trigger KVO notifications and which aren't. But then the `CHECK` lines went ahead and checked for the notifications that weren't supposed to be sent.

In other words, there were 3 lines that looked like

```swift
// CHECK-NEXT: swiftValue 13, objcValue four
```

but only 1 was supposed to be sent.

That said, from the test failure it sounds like that line is now printed zero times instead of the expected 1 time, so I guess it's still broken (and the implication being that `keyPathsAffectingValue(for:)` isn't being called for `\Target.objcValue`), but I haven't actually confirmed this yet (the alternative is that the test output is using zero-based line numbers instead of 1-based line numbers, which I suppose I'll find out whenever this build finishes).

kballard commented:
I've updated the PR to fix that test failure. The test itself is still broken, of course, but the current behavior now matches the previous one (as far as that test is concerned).

JunyuKuang commented:
Hi Lily, thank you for your contribution. Do you know any workaround for the **thread safety of KVO swizzling** crash? My app is randomly crash at launch with **Fatal error: Should never be reached** and it's really disturbing. I'm considering change all of my Swift KVO related implementation to Objective-C.

kballard commented:
@JunyuKuang There are only two workarounds for now:

1. Stop using Swift KVO. This is the most foolproof solution. I actually wrote and maintain an open-source library that does block-based KVO fast and thread-safe if you're interested, called [PMKVObserver](https://github.com/postmates/PMKVObserver). Or you could just use the existing old KVO registration method from Obj-C (you can still call it from Swift of course)
2. Ensure that you register a Swift KVO observer at some point prior to the use of KVO on a background thread. Pretty much the only way to actually guarantee this is to use an explicit `main.swift` and install the KVO prior to calling `NSApplicationMain(_:_:)`/`UIApplicationMain(_:_:_:_:)`. You can let the KVO uninstall itself immediately after installation, all that matters is that you have installed a KVO observer at least once prior to the use of KVO on a background thread.

lilyball committed:
Add some tests for KeyPath issues

Test to make sure the `String -> KeyPath` table used for the
`NSKeyValueObservingCustomization` callbacks isn't screwed up by
observing another object's property with the same name from within the
callbacks.

Test `NSSortDescriptor.keyPath` to ensure it's also not screwed up by
creating sort descriptors for the same property on different objects.

kballard commented:
I've added what tests I can reasonably write. I can't test the `NSKeyValueObservingCustomization` threading issues because of the global locks in the KVO implementation that prevent me from literally running the 2 `observe` calls simultaneously (and not literally running them simultaneously means I can't distinguish a threading issue from the other issue I am testing already, which is that we restore a prior keypath mapping after an observation).

This PR is ready to move forward.

kballard commented:
@millenomi Is there anything left for me to do on this PR?

millenomi commented:
@swift-ci please test

millenomi commented:
I'm still in the throes of https://github.com/apple/swift-corelibs-foundation/pull/1755 and won't have time to look at other stuff until it lands T_T

millenomi commented:
This is blocked by https://bugs.swift.org/browse/SR-9228.

zetasq commented:
Should we replace the initialize-once pattern below:

```
static var swizzler: ()? = {
    // Some operations
    return nil
}()
```

with

```
static let swizzler: Void = {
    // Some operations
}()
```

I think we don't need to return an optional void value here, and `let` should be used since it won't be mutated.

kballard commented:
Good point. It was written the way it is because it was adapted from a previous `static var swizzler : NSKeyValueObservation?` but you're right in that it should be perfectly fine to be a `let` and have the type `()`.

lilyball committed:
[SDK] Clean up the KVO swizzles slightly

kballard commented:
@millenomi [SR-9228](https://bugs.swift.org/browse/SR-9228) was merged 21 days ago. Is this still blocked?

millenomi commented:
It isn't; I still was v.v'

millenomi commented:
@swift-ci please test

millenomi commented:
Ah, dang. @kballard — this looks fine to me, but it needs merging.

millenomi commented on stdlib/public/SDK/Foundation/NSObject.swift:
```diff
@@ -127,65 +151,85 @@ fileprivate extension NSObject {
+            super.init()
+            objc_setAssociatedObject(object, associationKey(), self, .OBJC_ASSOCIATION_RETAIN)
+            __KVOKeyPathBridgeMachinery._withBridgeableKeyPath(from: path, to: keyPath) {
+                object.addObserver(self, forKeyPath: path, options: options, context: nil)
```
Do we want to have a private context for added safety?

kballard commented:
I can fix the conflicts, but the reported build failure is compilation failure in completely unrelated code.

millenomi commented:
Yeah; please ignore them — they probably just checked out your PR, erasing unrelated progress. (I didn't see the conflict marker until I asked the bot to start.)

Fixing the conflict should address that too.

kballard commented:
Rebasing my PR locally actually didn't produce any conflicts at all. I'm wondering if GitHub is reporting a conflict just because the number of commits between the branch and `master` (over 2000) is so high. The local rebase took a noticeable amount of time to do 3-way merges, so maybe GitHub doesn't want to do the work and just bails.

Anyway I'm just running a local build to sanity-check before updating the branch with the rebase.

kballard commented:
In fact, the "conflicts" are just that the `public/SDK` folder was renamed to `public/Darwin`. There was apparently no other changes to either file on `master` ¯\\\_(ツ)\_/¯

kballard commented:
Given that, I just updated the branch now instead of waiting for my local build.

millenomi commented:
@swift-ci please test

kballard commented:
```19:03:01 ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
```

millenomi commented:
:S let’s see if it was transient. (Heh.)

millenomi commented:
@swift-ci please test

millenomi commented:
@jrose-apple @Catfish-Man This patch is failing to compile in unrelated bridging code. Can you look at this? We want these fixes for 5.0.

jrose-apple commented:
I'm not sure why this is failing yet but you should be able to work around it by changing `init(_cocoaDictionary:)` to take an NSDictionary instead of an AnyObject. That'll be a tiny win for compile time too.

millenomi commented:
Please test with the following:
https://github.com/apple/swift/pull/21297

@swift-ci please test

jrose-apple commented:
:-( Unfortunately you can't do two PRs from the same repo with that functionality right now. (It's a good idea though! Worth filing a Radar about.)

millenomi commented:
urrrrghhhhhhh

millenomi commented:
If the fix PR tests positively, with your review @jrose-apple, I will land that and then resume the land sequence for this one.

millenomi commented:
rdar://46702942

millenomi commented:
@itaiferber @phausler that conditional looks yours

@jrose-apple it wasn't enough ;-;

itaiferber commented:
How did our tests pass but fail here?

jrose-apple commented:
@millenomi "And do the same thing for Set, and maybe anyone else using this AnyObject + `copy(with:)` pattrn, because I don't know what's causing it to suddenly fail with LilyB's changes." :-(

@itaiferber It's just a warning, because if you're checking for an architecture that doesn't exist, it's possible it's a real architecture that the current compiler just doesn't support.

millenomi commented:
@jrose-apple I need to figure out what that means on all other bridging code. Paging @Catfish-Man here as well because I'll need some direction in the bridging code.

jrose-apple commented:
I suspect there was a reason to use the opaque AnyObject in the past, but that there isn't one now. It should really have no effect.

millenomi commented:
Back from break, want this in, let's see how I can unblock it.

lilyball commented:
At this point, what's holding up this PR?

millenomi commented:
Let's see how the current status is first.

millenomi commented:
@swift-ci please test

millenomi commented:
@swift-ci please smoke test

millenomi commented:
So. With your permission, I'm going to commandeer this and iterate a little.

millenomi commented:
(that's to @lilyball.)

millenomi commented:
I'm trying a rebase; if the issue is still present, I will introduce a cast to `NSCopying` on those lines and benchmark.

millenomi commented:
Actually — first try @jrose-apple's thing and see if it introduces any ABI issue.

millenomi committed:
Rebase master; move this to use NSSet, not AnyObject.

millenomi commented:
@swift-ci please smoke test

millenomi committed:
The error is right: these should be @objc.

millenomi commented:
@swift-ci please smoke test

millenomi commented:
@swift-ci please smoke test

millenomi commented:
ok done

lilyball commented:
Thank you!

jrose-apple commented:
Looks like this is causing problems (on the simulator | on 32-bit platforms): https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RDA_test-simulator/1464/

(cc @shajrawi)

shajrawi commented:
there's a radar open for that - rdar://problem/47708852

millenomi commented:
Looking at it.

lilyball commented:
It sounds like the associated object trick for ensuring the observer is deregistered isn't working. Does the Obj-C 32-bit runtime check for leaked KVO observers happen before clearing associated objects?

I think the quick fix here is to just update the relevant test to ensure the `NSKeyValueObservation` is released prior to the `Dummy`. FWIW if the associated objects stuff isn't working, that just means it's behaving like it did prior to this PR (since the old code never even attempted that), and so the breakage is just because this PR modified the test.

lilyball commented:
I haven't tested it but I'd guess the following will work:

```diff
diff --git a/test/stdlib/KVOKeyPaths.swift b/test/stdlib/KVOKeyPaths.swift
index b3b302a367..32af1a4eac 100644
--- a/test/stdlib/KVOKeyPaths.swift
+++ b/test/stdlib/KVOKeyPaths.swift
@@ -127,13 +127,17 @@ class Target2 : NSObject, NSKeyValueObservingCustomization {
     // with the ability to look up the key path using the other.
     static func keyPathsAffectingValue(for key: AnyKeyPath) -> Set<AnyKeyPath> {
         print("keyPathsAffectingValue: key == \\.name:", key == \Target2.name)
-        _ = Dummy().observe(\.name) { (_, _) in }
+        withExtendedLifetime(Dummy()) { (dummy) in
+            _ = dummy.observe(\.name) { (_, _) in }
+        }
         return []
     }
 
     static func automaticallyNotifiesObservers(for key: AnyKeyPath) -> Bool {
         print("automaticallyNotifiesObservers: key == \\.name:", key == \Target2.name)
-        _ = Dummy().observe(\.name) { (_, _) in }
+        withExtendedLifetime(Dummy()) { (dummy) in
+            _ = dummy.observe(\.name) { (_, _) in }
+        }
         return true
     }
 }
```

millenomi commented:
I need to check if automatic reregistration works on 32-bit.

millenomi commented:
@lilyball your suggestion works, and I will repropose your patch with the fix.

rob-keepsafe commented:
Just an FYI for others: this fix still isn't released as of iOS 12.2 / Xcode 10.2.1, making Swift KVO unusable on background threads for now.

bobergj commented:
@CLoutas: git branch -r --contains 7c514196fcef1e59673a46c6a452c38d9322cf34 shows this was merged into the branch "swift-5.1-branch". Meaning it was included in Swift 5.1.

Since the fixes affect the runtime libraries rather than the compiler, this means that it is fixed on iOS 13, macOS 10.15 but still requires the workaround on earlier OS versions.

CLoutas commented:
@bobergj that's unfortunate, but at least we have a workaround. Thank you very much for clarifying that.
